### PR TITLE
Protocol and metamodel changes for Mastery Runtime

### PR DESCRIPTION
**Pull request by agshubh-97**
## How did you test this change?

- [ ] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

**Comment by agshubh-97** [comment]
Original PR: https://github.com/finos/legend-studio/pull/2677

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 1]
```diff
@@ -257,6 +269,41 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
...
+          for (const builder of extraMasteryRuntimeBuilders) {
+            const metamodel = builder(
+              elementProtocol as V1_MasteryRuntime,
+              context,
+            );
+            if (metamodel) {
+              return metamodel;
+            }
+          }
+          throw new UnsupportedOperationError(
```
Comment (agshubh-97) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts:295]: @gayathrir11 I tried to add a `return undefined` here but was getting Lint Issues as the return type for this function is `PackageableElement` 
Can you please suggest the workaround?

Comment (gayathrir11) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts:295]: The error you are throwing here should be fine as we expect to return a metamodel if we find any protocol of type `V1_Masteryruntime`

Comment (agshubh-97) [reply packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts:295]: Just wondering, what about non Alloy implementations of Legend Studio. Will it be necessary for a user to implement a protocol of type V1_MasteryRuntime? Is that fine?

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 2]
```diff
@@ -257,6 +266,41 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
...
+            `Can't build runtime '${typeof elementProtocol}'`,
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts:293]: Can you please modify this error message to `Can't build mastery runtime '${typeof elementProtocol}'`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts [hunk 3]
```diff
@@ -393,4 +444,23 @@ export class DSL_Mastery_PureProtocolProcessorPlugin
       },
     ];
   }
+
+  V1_callExtraDeserializers = (
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin.ts:448]: Can you please rename this to `V1_deserializeMasteryRuntime`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts [hunk 4]
```diff
@@ -819,3 +821,28 @@ export const V1_transformMasterRecordDefinition = (
...
+      ).V1_getExtraMasteryRuntimeTransformers?.() ?? [],
+  );
+  for (const transformer of extraMasteryRuntimeTransformers) {
+    const protocol = transformer(element, context);
+    if (protocol) {
+      return protocol;
+    }
+  }
+  throw new UnsupportedOperationError(
+    `Can't transform runtime '${typeof element}'`,
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_DSL_Mastery_TransformerHelper.ts:846]: Please modify error message to: `Can't transform mastery runtime: : no compatible transformer available from plugins, element`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/to/V1_DSL_Mastery_BuilderHelper.ts [hunk 5]
```diff
@@ -830,3 +832,28 @@ export const V1_buildMasterRecordDefinition = (
...
+      ).V1_getExtraMasteryRuntimeBuilders?.() ?? [],
+  );
+  for (const builder of extraMasteryRuntimeBuilders) {
+    const metamodel = builder(element, context);
+    if (metamodel) {
+      return metamodel;
+    }
+  }
+  throw new UnsupportedOperationError(
+    `Can't build runtime '${typeof element}'`,
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureGraph/to/V1_DSL_Mastery_BuilderHelper.ts:857]: Can you update error message similar to i commenetd above

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin_Extension.ts [hunk 6]
```diff
@@ -169,4 +192,15 @@ export interface DSL_Mastery_PureProtocolProcessorPlugin_Extension
   V1_getExtraAuthenticationStrategyProtocolSerializers?(): V1_AuthenticationStrategyProtocolSerializer[];
 
   V1_getExtraAuthenticationStrategyProtocolDeserializers?(): V1_AuthenticationStrategyProtocolDeserializer[];
+
+  // extension hooks: runtime
+  V1_getExtraMasteryRuntimeBuilders?(): V1_MasteryRuntimeBuilder[];
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/DSL_Mastery_PureProtocolProcessorPlugin_Extension.ts:197]: Can you please rename this to `V1_getExtraMasteryRuntimeSecondPassBuilders`

File: packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts [hunk 7]
```diff
@@ -926,3 +927,52 @@ export const V1_masterRecordDefinitionModelSchema = (
...
+      return masteryRuntimeJson;
+    }
+  }
+  throw new UnsupportedOperationError(
+    `Can't serialize mastery runtime: no compatible serializer available from plugins`,
+    protocol,
+  );
+};
+
+export const V1_deserializeMasteryRuntime = (
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-mastery/src/graph-manager/protocol/pure/v1/transformation/pureProtocol/V1_DSL_Mastery_ProtocolHelper.ts:959]: Are we using it anywhere?

**Review by gayathrir11 — CHANGES_REQUESTED** [review]

**Review by gayathrir11 — APPROVED** [review]

**Pull request merged**